src/cdk-experimental/popover-edit/edit-event-dispatcher.ts

```diff
@@ -37,13 +68,130 @@ export class EditEventDispatcher {
   }
   private _editRef: EditRef<any>|null = null;
 
+  // Optimization: Precompute common pipeable operators used per row/cell.
+  private readonly _distinctUntilChanged =
+      distinctUntilChanged<Element|HoverContentState|boolean|null>();
+  private readonly _startWithNull = startWith<Element|null>(null);
+  private readonly _distinctShare = pipe(
+    this._distinctUntilChanged as MonoTypeOperatorFunction<HoverContentState>,
+    share(),
+  );
+  private readonly _startWithNullDistinct = pipe(
+    this._startWithNull,
+    this._distinctUntilChanged as MonoTypeOperatorFunction<Element|null>,
+  );
+
+  /** An observable that emits the row containing focus or an active edit. */
+  readonly editingOrFocused = combineLatest(
+      this.editing.pipe(
+          map(cell => closest(cell, ROW_SELECTOR)),
+          this._startWithNull,
+      ),
+      this.focused.pipe(this._startWithNull),
+  ).pipe(
+      map(([editingRow, focusedRow]) => focusedRow || editingRow),
+      this._distinctUntilChanged as MonoTypeOperatorFunction<Element|null>,
+      auditTime(FOCUS_DELAY), // Use audit to skip over blur events to the next focused element.
+      this._distinctUntilChanged as MonoTypeOperatorFunction<Element|null>,
+      share(),
+  );
+
+  /** Tracks rows that contain hover content with a reference count. */
+  private _rowsWithHoverContent = new Map<Element|null, number>();
+
   /** The table cell that has an active edit lens (or null). */
   private _currentlyEditing: Element|null = null;
 
-  private readonly _hoveringDistinct = this.hovering.pipe(distinctUntilChanged(), share());
-  private readonly _editingDistinct = this.editing.pipe(distinctUntilChanged(), share());
+  /** The combined set of row hover content states organized by row. */
+  private readonly _hoveredContentStateDistinct = combineLatest(
+      this.allRows.pipe(
```

Comment:
Could you add a comment that explains the loop forward/backwards first-row/last-row thing?